runtime/js/40_testing.js

```diff
@@ -1408,6 +1424,12 @@
     return value == null ? defaultValue : value;
   }
 
+  let teardownFn;
```

Comment:
Do we need to consider that `test.teardown` might be called multiple times? This could lead to some very unexpected and hard to debug behaviour with the current implementation.

Reply:
I think we need to consider that case. Is there any good idea for handling that?

Reply:
how about this? push `teardownFn` and call sequentially after the test. 
```typescript
const teardownFns = [];
function teardown(fn) {
	if(!teardownFns.includes(fn)) teardownFns.push(fn);
}

// after the test
for await (const fn of teardownFns) {
	await fn();
}
```
but I'm not sure this is good about the order of calls.
To handle the order problem, we can take another approach.
```typescript
let teardownFn = () => {};
function teardown(fn) {
	teardownFn = fn(teardownFn);
}

// example of fn
function exampleFn(alreadyRegisteredFn) {
	return async function () {
		// await alreadyRegisteredFn();
		doAwesomeStuff();
		// await alreadyRegisteredFn(); <- we can choose the call order of `teardownFns`.
	}
}
```

Reply:
This is a good point!

I think the second example you gave could be improved in the following ways:

 1. It is hard to understand the order of calls because developers must check each `teardown` function to know what the order is.
 2. It is prone to errors e.g. a developer forgets to call the previous `teardown` function or calls it twice

The issue is that each test has a degree of control over the order of `teardown` calls, essentially democratising the order. This approach probably makes problem `#1` unsolvable.

How important is it that developers have control over the order of calls? If fine grained control over the order is important, then it will be possible to implement a custom solution:

```ts
function doTeardownOperationsInSomeOrder() {
  // --snip--
}

Deno.test("name", (test) => {
  test.teardown(doTeardownOperationsInSomeOrder);

  test.step("stepName", (test) => {
    // Custom solution to register teardown operations to be called by `doTeardownOperationsInSomeOrder`
  }

  // --snip--
});
```

However, I believe that in 99% of cases this problem would be solved by Deno exposing the following functions in the test context:

```ts
Deno.test("test", (test) => {
  test.beforeAll(() => { /* --snip-- */ });
  test.beforeEach(() => { /* --snip-- */ });
  test.afterAll(() => { /* --snip-- */ });
  test.afterEach(() => { /* --snip-- */ });
});
```

These would work similarly to the [Jest Setup and Teardown functions](https://jestjs.io/docs/setup-teardown) and allow for a decent level of control over the order while still being easy to understand.

To summarise, I propose the following:

**test.beforeAll** - _[NOT IN THIS PR]_ execute a function before any child steps are executed
**test.afterAll** - _[NOT IN THIS PR]_ execute a function after any child steps are executed
**test.beforeEach** - _[NOT IN THIS PR]_ execute a function before each child steps is executed
**test.afterEach** - _[NOT IN THIS PR]_ execute a function after each child steps is executed
**test.teardown** - _[THIS PR]_ execute a function after the top level parent test has finished executing and all descendant test steps have finished executing

What do you think @hyp3rflow?

Reply:
Thanks for the kind review!
Yeah, I think it will be okay with my first suggestion in most cases. I was just worried about if there are many third-party libraries using teardown in an order-sensitive way.
I also agree that it will be awesome with APIs like jest setup-teardown functions. but I think we need to hear the maintainers' thoughts about this kind of API.

Reply:
> but I think we need to hear the maintainers' thoughts about this kind of API.

Definitely agree! And this should not be part of this PR either way. I just mentioned it here as it would address the ordering issue you flagged.

> I was just worried about if there are many third-party libraries using teardown in an order-sensitive way.

I am not aware of any but it's a valid point.

> Yeah, I think it will be okay with my first suggestion in most cases.

I agree! Better to keep it simple :D

Reply:
@bcheidemann I have a question.
> test.teardown - [THIS PR] execute a function after the top level parent test has finished executing and all descendant test steps have finished executing. 

What does `top level parent test` means? A single deno test (Deno.test()) or entire test module (a.test.ts)?

Reply:
> @bcheidemann I have a question.
> 
> > test.teardown - [THIS PR] execute a function after the top level parent test has finished executing and all descendant test steps have finished executing.
> 
> What does `top level parent test` means? A single deno test (Deno.test()) or single test module (a.test.ts)?

I was thinking a single `Deno.test()` but what do you think is better?

For our case (`assertSnapshot`) it probably makes more sense for it to be a single test module (`a.test.ts`) but I think generally it makes more sense for this to relate to a single `Deno.test()` instance.